### PR TITLE
Forms/select

### PR DESCRIPTION
**Pull request by dprokop**
This PR introduces revamped Select component. 

### Approach

First I've looked into current Select implementation to check if there is an easy way to wrap it and customise to our new design needs. Unfortunately, there are couple of things that blocked me from wrapping the current Select:
- current implementation heavily relies on gf-form class names
- it also utilises Tooltip which i believe should not belong to the Select component. If tooltip is necessary for a Select, then composition should be used

I took approach to slightly rewrite it while still using the react-select under the hood.

So, the new Forms.Select has a similar API to the current Select with some small implementations simplified and API adjustments:

- There is `BaseSelect` component which renders Select, Creatable or ReactAsyncSelect select depending on the props passed. This component is used to implement [Select](https://github.com/grafana/grafana/pull/21092/files#diff-cad90178586f6ddfae6ba2cad3f6d738R287) and [AsyncSelect](https://github.com/grafana/grafana/pull/21092/files#diff-cad90178586f6ddfae6ba2cad3f6d738R293) which implementation is way simpler now
- Select API has now a property `renderControl` responsible for rendering the select trigger. This simplified the implementation of [ButtonSelect](https://github.com/grafana/grafana/pull/21092/files#diff-d8917983e359de11482175f22e90e073R14)
- `loadingMessage` and `noOptionsMessage` props accept string instead of function, as we don't use the dynamic messages(based on current input value) anywhere
- `isDisabled` prop is now renamed to `disabled`
- API interfaces are moved to the new Select implementation. The "old" select uses the new interfaces and adjusts them for backwards compatibility (https://github.com/grafana/grafana/pull/21092/files#diff-e81797cd97fef7c9269b07eae8290718R32). This is primarily overriding the adjusted props mentioned above
- New Select (overrides the components of react-select)[https://github.com/grafana/grafana/pull/21092/files#diff-cad90178586f6ddfae6ba2cad3f6d738R221] which is used for applying the revamped styles.
- The `width` property (is deprecated)[https://github.com/grafana/grafana/pull/21092/files#diff-cad90178586f6ddfae6ba2cad3f6d738R200] in favor of (predefined input sizing)[https://github.com/grafana/grafana/pull/21092/files#diff-cad90178586f6ddfae6ba2cad3f6d738R264]. Custom width should be provided via `className`

TODO:
- [x] When value is selected the focus is stuck on the input even when user clicks outside of it
- [x] Adjust `value` interface to accept not only `SelectableValue<T>` but also `T`
- [x] `ButtonSelect` - align with current implementation 
- [x] Add tests
- [ ] Keyboard navigation with custom scrollbar (followup issue)

### CODE REVIEW

File: packages/grafana-ui/package.json [hunk 1]
```diff
@@ -29,6 +29,7 @@
     "@grafana/slate-react": "0.22.9-grafana",
     "@torkelo/react-select": "2.1.1",
     "@types/react-color": "2.17.0",
+    "@types/react-select": "^2.0.15",
```
Comment (dprokop) [inline packages/grafana-ui/package.json]: Needs fixing

File: packages/grafana-ui/src/components/Forms/Select/Select.tsx [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+  return (
+    <Component
+      components={{
+        MenuList: SelectMenu,
+        Group: SelectOptionGroup,
+        ValueContainer: ValueContainer,
+        IndicatorsContainer: IndicatorsContainer,
+        IndicatorSeparator: () => <></>,
+        Control: getControlComponent(prefix, disabled, invalid, renderControl),
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/Select.tsx]: Control element can be customised with `renderControl` prop

Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/Select.tsx]: Control element can be customised with `renderControl` prop

File: packages/grafana-ui/src/components/Select/Select.tsx [hunk 3]
```diff
@@ -14,54 +14,34 @@ import { components } from '@torkelo/react-select';
...
-  openMenuOnFocus?: boolean;
-  onBlur?: () => void;
-  maxMenuHeight?: number;
-  isLoading?: boolean;
+/**
+ * Changes in new selects:
+ * - noOptionsMessage & loadingMessage is of string type
+ * - isDisabled is renamed to disabled
+ */
+type LegacyCommonProps<T> = Omit<SelectCommonProps<T>, 'noOptionsMessage' | 'disabled'>;
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Select/Select.tsx]: Old select uses the same interface as the new one. But there are minor changes in the new Select interface, so I'm making sure nothing breaks in the old select

**Review by torkelo — COMMENTED** [review]
This is looking great! and looks like a lot of work, not strange that this took a few days!

**Review by peterholmberg — COMMENTED** [review]
What is the need to have Forms components with forwardRef type?

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  maxHeight: number;
+}
+
+export const SelectMenu = React.forwardRef<HTMLDivElement, React.PropsWithChildren<SelectMenuProps>>((props, ref) => {
+  const theme = useTheme();
+  const styles = getSelectStyles(theme);
+  // @ts-ignore
+  const { children, maxHeight, innerRef } = props;
+
+  console.log('SELECT MENU', props);
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]:
```suggestion
```

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  renderOptionLabel?: (value: SelectableValue<T>) => JSX.Element;
+  data: SelectableValue<T>;
+}
+
+export const SelectMenuOptions = React.forwardRef<HTMLDivElement, React.PropsWithChildren<SelectMenuOptionProps<any>>>(
+  (props, ref) => {
+    const theme = useTheme();
+    const styles = getSelectStyles(theme);
+    const { children, innerProps, data, renderOptionLabel, isSelected, isFocused } = props;
+    console.log('menu options', innerProps);
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]:
```suggestion
```

File: packages/grafana-ui/src/components/Forms/Select/Select.story.tsx [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+export const multiPlainValue = () => {
+  const [value, setValue] = useState<string[]>();
+
+  return (
+    <>
+      <MultiSelect
+        options={generateOptions()}
+        value={value}
+        onChange={v => {
+          console.log(v);
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/Select.story.tsx]:
```suggestion
```

File: packages/grafana-ui/src/components/Forms/Select/Select.story.tsx [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+export const multiSelectAsync = () => {
+  const [value, setValue] = useState<Array<SelectableValue<string>>>();
+
+  return (
+    <AsyncMultiSelect
+      loadOptions={loadAsyncOptions}
+      defaultOptions
+      value={value}
+      onChange={v => {
+        console.log(v);
```
Comment (dprokop) [inline packages/grafana-ui/src/components/Forms/Select/Select.story.tsx]:
```suggestion
```

**Review by torkelo — APPROVED** [review]
Did some testing to make sure there are no regressions and found none! 

Great job on this one!

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+// @ts-ignore
+import { components } from '@torkelo/react-select';
+interface SelectMenuProps {
+  maxHeight: number;
+}
+
+export const SelectMenu = React.forwardRef<HTMLDivElement, React.PropsWithChildren<SelectMenuProps>>((props, ref) => {
+  const theme = useTheme();
+  const styles = getSelectStyles(theme);
+  // @ts-ignore
```
Comment (peterholmberg) [inline packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]: And this ts-ignore?

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import { useTheme } from '../../../themes/ThemeContext';
+import { getSelectStyles } from './getSelectStyles';
+import { cx } from 'emotion';
+import { SelectableValue } from '@grafana/data';
+import { Icon } from '../../Icon/Icon';
+import { CustomScrollbar } from '../../CustomScrollbar/CustomScrollbar';
+// @ts-ignore
```
Comment (peterholmberg) [inline packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]: and this?

**Review by peterholmberg — CHANGES_REQUESTED** [review]
Left some comments and questions.

File: packages/grafana-ui/src/components/Forms/Select/Select.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { SelectableValue } from '@grafana/data';
+import { SelectCommonProps, SelectBase, MultiSelectCommonProps, SelectAsyncProps } from './SelectBase';
+
+export function Select<T>(props: SelectCommonProps<T>) {
+  return <SelectBase {...props} />;
+}
+
+export function MultiSelect<T>(props: MultiSelectCommonProps<T>) {
+  // @ts-ignore
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Select/Select.tsx]: MultiSelect onChange property is different (accepts array) than SelectBase. It's overridden in the MultiSelectCommonProps type, but the base type of SelectBase still has the onChange definition for a single select. Hence the ts-ignore

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx (hunk 8 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]: innerRef is not typed, I'll fix it

File: packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx (hunk 9 above)
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Select/SelectMenu.tsx]: to be removed

File: packages/grafana-ui/src/components/Forms/Select/Select.story.tsx [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+
+export default {
+  title: 'UI/Forms/Select',
+  component: Select,
+  decorators: [withCenteredStory, withHorizontallyCenteredStory],
+  // parameters: {
+  //   docs: {
+  //     page: mdx,
+  //   },
+  // },
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Select/Select.story.tsx]: It's by default

File: packages/grafana-ui/src/components/Forms/Select/SelectBase.tsx [hunk 12]
```diff
@@ -0,0 +1,308 @@
+import React from 'react';
+import { SelectableValue, deprecationWarning } from '@grafana/data';
+// @ts-ignore
+import { default as ReactSelect, Creatable, components } from '@torkelo/react-select';
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Select/SelectBase.tsx]: yes

File: packages/grafana-ui/src/components/Forms/Button.story.tsx [hunk 13]
```diff
@@ -1,6 +1,8 @@
 import React from 'react';
 import { Button, ButtonVariant } from './Button';
 import { withCenteredStory, withHorizontallyCenteredStory } from '../../utils/storybook/withCenteredStory';
+import { getIconKnob } from '../../utils/storybook/knobs';
+
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-ui/src/components/Forms/Button.story.tsx]: yes

**Comment by ryantxu** [comment]
Not sure this is the right PR/place... the other feature on select that would be great is changing the order of multi-select tags.  See:
https://react-select.com/advanced

File: packages/grafana-ui/src/components/Forms/Select/Select.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (peterholmberg) [reply packages/grafana-ui/src/components/Forms/Select/Select.tsx]: Ok

**Review by peterholmberg — APPROVED** [review]
Looks good now! :D

File: packages/grafana-ui/src/components/Button/Button.tsx [hunk 14]
```diff
@@ -32,14 +33,14 @@ export const Button: React.FunctionComponent<ButtonProps> = props => {
       variant: variant || 'primary',
     });
 
+  const buttonClassName = cx(styles.button, icon && styles.buttonWithIcon, icon && !children && styles.iconButton);
   return (
-    <button className={cx(styles.button, className)} {...buttonProps}>
```
Comment (alexanderzobnin) [inline packages/grafana-ui/src/components/Button/Button.tsx]: `className` isn't passed into the button anymore. Is it a bug or feature?

**Pull request merged**